test/functional/attachments_controller_test.rb

```diff
@@ -1,188 +1,473 @@
+    assert_redirected_to view_context.path_to_image('thumbnail-placeholder.png')
   end
 
-  def create_thumbnail_for_upload(uploader)
-    FileUtils.touch("#{uploader.clean_path}.png")
+  test 'redirects to placeholder page if file is unscanned non-image' do
+    setup_stubs(file_state: :unscanned)
+
+    get :show, params: params
+
+    assert_response :found
+    assert_redirected_to placeholder_url
   end
 
-  test "requesting an attachment's thumbnail returns the thumbnail inline" do
-    attachment_data = build(:attachment_data)
-    attachment = build(:file_attachment, attachment_data: attachment_data)
+  test 'sets Cache-Control header max-age & public directives if unscanned non-image' do
+    setup_stubs(file_state: :unscanned)
 
-    create(:published_publication, :with_file_attachment, attachments: [attachment])
-    VirusScanHelpers.simulate_virus_scan(attachment_data.file)
-    create_thumbnail_for_upload(attachment_data.file)
-    get :show, params: { id: attachment_data.to_param, file: attachment_data.filename, extension: 'png' }
+    get :show, params: params
 
-    assert_response :success
-    assert_match "#{attachment_data.filename}.png", response.headers['Content-Disposition']
-    assert_match %r[^inline;], response.headers['Content-Disposition']
+    assert_cache_control 'max-age=60'
+    assert_cache_control 'public'
   end
 
-  test 'requesting an attachment that has not been virus checked redirects to the placeholder page' do
-    attachment_data = build(:attachment_data)
-    attachment = build(:file_attachment, attachment_data: attachment_data)
+  test 'redirects to placeholder page if file is unscanned non-image even if deleted' do
+    setup_stubs(file_state: :unscanned, deleted?: true)
+
+    get :show, params: params
 
-    create(:published_publication, :with_file_attachment_not_scanned, attachments: [attachment])
+    assert_response :found
+    assert_redirected_to placeholder_url
+  end
+
+  test 'redirects to placeholder page if file is unscanned non-image even if draft & not accessible' do
+    setup_stubs(file_state: :unscanned, draft?: true, accessible_to?: false)
 
-    get_show attachment_data
+    get :show, params: params
 
+    assert_response :found
     assert_redirected_to placeholder_url
-    assert_cache_control "max-age=#{1.minute}"
   end
 
-  test "requesting an attachment on an unpublished edition redirects to the edition's unpublishing page" do
-    unpublished_edition = create(:draft_publication, :unpublished, :with_file_attachment)
-    attachment_data = unpublished_edition.attachments.first.attachment_data
+  # Not found
 
-    get_show attachment_data
+  test 'responds with 404 Not Found if attachment data does not exist' do
+    setup_stubs
+    controller.stubs(:attachment_data).raises(ActiveRecord::RecordNotFound)
 
-    assert_redirected_to publication_url(unpublished_edition.unpublishing.slug)
+    assert_raises(ActiveRecord::RecordNotFound) { get :show, params: params }
```

Comment:
Could you use `assert_response :not_found` here too?

Reply:
Unfortunately I don't think so. The problem is that the middleware (?) that converts a `ActiveRecord::RecordNotFound` into a `404 Not Found` is not used in the controller tests.